### ISSUE TITLE
`torch.compile(optmizer.step)` with lr scheduler and tensor lr still cause a recompilation in the second iteration

### ISSUE THREAD

**Issue by akihironitta** [pytorch/pytorch#130467]
### 🐛 Describe the bug

I see `torch.compile` still recompiles `optimizer.step` in the second iteration when using a learning rate scheduler as produced in the following script (or [here](https://github.com/akihironitta/gist/tree/repro/torch-guard-on-initial-lr/pl_boring_model)).

\c\c @mlazos as the author of https://pytorch.org/tutorials/recipes/compiling_optimizer_lr_scheduler.html

### Error logs

```console
$ TORCH_LOGS=recompiles python pl_boring_model/main.py
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
/home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:75: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `pytorch_lightning` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
/home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/configuration_validator.py:70: You defined a `validation_step` but have no `val_dataloader`. Skipping val loop.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:424: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=15` in the `DataLoader` to improve performance.
/home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/loops/fit_loop.py:298: The number of training batches (8) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.
W0710 18:14:13.417000 140121105339008 torch/_logging/_internal.py:1013] [1/0] Profiler function <class 'torch.autograd.profiler.record_function'> will be ignored
V0710 18:14:16.882000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles] Recompiling function _call_strategy_hook in /home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/call.py:294
V0710 18:14:16.882000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles]     triggered by the following guard failure(s):
V0710 18:14:16.882000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles]     - L['hook_name'] == 'training_step'
V0710 18:14:16.891000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles] Recompiling function torch_dynamo_resume_in__call_strategy_hook_at_300 in /home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/call.py:300
V0710 18:14:16.891000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles]     triggered by the following guard failure(s):
V0710 18:14:16.891000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles]     - L['hook_name'] == 'training_step'
V0710 18:14:19.659000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles] Recompiling function torch_dynamo_resume_in_step_at_148 in /home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/torch/optim/adam.py:148
V0710 18:14:19.659000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles]     triggered by the following guard failure(s):
V0710 18:14:19.659000 140121105339008 torch/_dynamo/guards.py:1425] [__recompiles]     - L['self'].param_groups[0]['lr'] is L['self'].param_groups[0]['initial_lr']
`Trainer.fit` stopped: `max_epochs=5` reached.
```

### Minified repro

```python
# pl_boring_model/main.py
from pytorch_lightning import LightningModule, Trainer
import torch
from torch.utils.data import DataLoader, Dataset


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        return self(batch).sum()

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.layer.parameters(), lr=torch.tensor(0.1))
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        optimizer.step = torch.compile(optimizer.step)
        return [optimizer], [lr_scheduler]


def main():
    train_data = DataLoader(RandomDataset(32, 16), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        max_epochs=5,
        enable_checkpointing=False,
        enable_model_summary=False,
        enable_progress_bar=False,
    )
    trainer.fit(model, train_dataloaders=train_data)


if __name__ == "__main__":
    main()
```

### Versions

```
Collecting environment information...
PyTorch version: 2.3.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.31

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-1055-aws-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 545.23.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             16
On-line CPU(s) list:                0-15
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
Stepping:                           7
CPU MHz:                            2499.998
BogoMIPS:                           4999.99
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          256 KiB
L1i cache:                          256 KiB
L2 cache:                           8 MiB
L3 cache:                           35.8 MiB
NUMA node0 CPU(s):                  0-15
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke avx512_vnni

Versions of relevant libraries:
[pip3] numpy==2.0.0
[pip3] pytorch-lightning==2.3.3
[pip3] torch==2.3.1
[pip3] torchmetrics==1.4.0.post0
[pip3] triton==2.3.1
[conda] numpy                     2.0.0                    pypi_0    pypi
[conda] pytorch-lightning         2.3.3                    pypi_0    pypi
[conda] torch                     2.3.1                    pypi_0    pypi
[conda] torchmetrics              1.4.0.post0              pypi_0    pypi
[conda] triton                    2.3.1                    pypi_0    pypi
```

cc @ezyang @anijain2305 @chauhang @penguinwu

**Comment by akihironitta** [comment]
Tried 2.4.0rc, and it no longer recompiles `optimizer.step` in the second iteration:
```console
$ TORCH_LOGS=recompiles python pl_boring_model/main.py
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
/home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:75: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `pytorch_lightning` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
/home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/configuration_validator.py:70: You defined a `validation_step` but have no `val_dataloader`. Skipping val loop.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:424: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=15` in the `DataLoader` to improve performance.
/home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/loops/fit_loop.py:298: The number of training batches (8) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.
W0710 18:28:01.050000 139780505379456 torch/_logging/_internal.py:1034] [1/0] Profiler function <class 'torch.autograd.profiler.record_function'> will be ignored
V0710 18:28:04.279000 139780505379456 torch/_dynamo/guards.py:2611] [4/1] [__recompiles] Recompiling function _call_strategy_hook in /home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/call.py:294
V0710 18:28:04.279000 139780505379456 torch/_dynamo/guards.py:2611] [4/1] [__recompiles]     triggered by the following guard failure(s):
V0710 18:28:04.279000 139780505379456 torch/_dynamo/guards.py:2611] [4/1] [__recompiles]     - L['hook_name'] == 'training_step'
V0710 18:28:04.286000 139780505379456 torch/_dynamo/guards.py:2611] [5/1] [__recompiles] Recompiling function torch_dynamo_resume_in__call_strategy_hook_at_300 in /home/aki/miniconda3/envs/oss310/lib/python3.10/site-packages/pytorch_lightning/trainer/call.py:300
V0710 18:28:04.286000 139780505379456 torch/_dynamo/guards.py:2611] [5/1] [__recompiles]     triggered by the following guard failure(s):
V0710 18:28:04.286000 139780505379456 torch/_dynamo/guards.py:2611] [5/1] [__recompiles]     - L['hook_name'] == 'training_step'
```

Sorry for making the noise here, and thank you for making PyTorch greater! ❤️
